File: legacy/engine/context.py
"""Ctx — the memoize-or-execute walk, run inside Temporal workflow code.

Workflow code is deterministic: every DB / IO touch happens in an activity.
Per ctx.node call:
    input_hash = H(canonical argument map)      (artifacts by content hash)
    memo_key   = H(code_hash || input_hash)
    lookup (engagement_id, memo_key) -> hit: reuse; miss: execute (engine
    activity, or human-task workflow + poll) -> idempotent completion tx.
"""

from __future__ import annotations

import json
from datetime import timedelta
from typing import Any

from temporalio import workflow
from temporalio.common import RetryPolicy
from temporalio.exceptions import ApplicationError

from engine.canonical import hash_value, memo_key as make_memo_key

_ENGINE_RETRY = RetryPolicy(maximum_attempts=5, non_retryable_error_types=["NodeError"])
_SHORT = timedelta(seconds=30)
_NODE_TIMEOUT = timedelta(seconds=120)


class ArtifactHandle:
    """Immutable handle to an artifact. In workflow code it is reference-only;
    inside node bodies (activities) it gains payload access via a loader."""

    __slots__ = ("ref", "_loader")

    def __init__(self, ref: dict, loader=None):
        self.ref = ref
        self._loader = loader

    @property
    def artifact_id(self) -> int:
        return self.ref["artifact_id"]

    @property
    def hash(self) -> str:
        return self.ref["hash"]

    @property
    def kind(self) -> str:
        return self.ref["kind"]

    @property
    def label(self) -> str:
        return self.ref.get("label") or ""

    @property
    def media_type(self) -> str:
        return self.ref.get("media_type") or "application/octet-stream"

    def bytes(self) -> bytes:
        if self._loader is None:
            raise RuntimeError(
                "payload access is only legal inside node bodies (activities), "
                "never in workflow code"
            )
        return self._loader(self.artifact_id)

    def text(self) -> str:
        return self.bytes().decode("utf-8")

    def json(self) -> Any:
        return json.loads(self.text())

    def __repr__(self) -> str:  # pragma: no cover
        return f"ArtifactHandle({self.kind}#{self.artifact_id} {self.hash[:8]})"


def _encode(value: Any) -> tuple[Any, Any, list[int]]:
    """-> (hash_form, transport_form, input_artifact_ids).
    hash_form: artifacts as {"$artifact": hash}; artifact lists sorted by hash.
    transport_form: artifacts tagged {"__artifact__": ref} for the activity."""
    if isinstance(value, ArtifactHandle):
        return {"$artifact": value.hash}, {"__artifact__": dict(value.ref)}, [value.artifact_id]
    if isinstance(value, (list, tuple)):
        items = list(value)
        if items and all(isinstance(v, ArtifactHandle) for v in items):
            items = sorted(items, key=lambda h: h.hash)  # canonical rule 6
        hashes, transports, ids = [], [], []
        for v in items:
            h, t, i = _encode(v)
            hashes.append(h)
            transports.append(t)
            ids.extend(i)
        return hashes, transports, ids
    if isinstance(value, dict):
        hashes, transports, ids = {}, {}, []
        for k, v in value.items():
            h, t, i = _encode(v)
            hashes[k] = h
            transports[k] = t
            ids.extend(i)
        return hashes, transports, ids
    return value, value, []


class Ctx:
    def __init__(self, inp: dict):
        self.engagement_id: int = inp["engagement_id"]
        self.workflow_run_id: int = inp["workflow_run_id"]
        self.workflow_id: str = inp["workflow_id"]
        self._declared: set[str] = set(inp["declared_kinds"])
        self._attachments = [ArtifactHandle(r) for r in inp["attachments"]]
        self.executed: list[str] = []
        self.memo_hits: list[str] = []
        self.human_waits: list[str] = []

    # ---------- attachment resolution (user-sourced snapshot only, I7) ----------

    def attached(self, kind: str) -> list[ArtifactHandle]:
        if kind not in self._declared:
            raise ApplicationError(
                f"kind {kind!r} is not declared by workflow {self.workflow_id!r}",
                non_retryable=True,
            )
        return sorted(
            (h for h in self._attachments if h.kind == kind), key=lambda h: h.hash
        )

    def attached_one(self, kind: str) -> ArtifactHandle:
        items = self.attached(kind)
        if len(items) != 1:
            raise ApplicationError(
                f"expected exactly one {kind!r} attachment, found {len(items)}",
                non_retryable=True,
            )
        return items[0]

    def attached_one_or_none(self, kind: str) -> ArtifactHandle | None:
        items = self.attached(kind)
        if len(items) > 1:
            raise ApplicationError(
                f"expected at most one {kind!r} attachment, found {len(items)}",
                non_retryable=True,
            )
        return items[0] if items else None

    # pure aliases: named for the override pattern
    user_supplied = attached
    user_supplied_one = attached_one
    user_supplied_one_or_none = attached_one_or_none

    # ---------- the walk ----------

    async def node(self, fn, *args: Any, **kwargs: Any) -> ArtifactHandle:
        nd = getattr(fn, "_engine_node", None)
        if nd is None:
            raise ApplicationError(
                f"{fn!r} is not a registered node (missing @node/@human_node)",
                non_retryable=True,
            )
        arg_map: dict[str, Any] = dict(zip(nd.param_names, args))
        for k, v in kwargs.items():
            if k not in nd.param_names:
                raise ApplicationError(
                    f"unknown parameter {k!r} for node {nd.node_id}", non_retryable=True
                )
            arg_map[k] = v
        for p in nd.param_names:
            arg_map.setdefault(p, None)  # absent optional input: explicit null (rule 7)

        hash_form, transport, input_ids = _encode(arg_map)
        input_hash = hash_value(hash_form)
        mk = make_memo_key(nd.code_hash, input_hash)

        ref = await workflow.execute_activity(
            "memo_lookup",
            args=[self.engagement_id, mk],
            start_to_close_timeout=_SHORT,
        )
        if ref is not None:
            self.memo_hits.append(nd.node_id)
            await self._attach(ref)
            return ArtifactHandle(ref)

        req = {
            "engagement_id": self.engagement_id,
            "workflow_run_id": self.workflow_run_id,
            "workflow_id": self.workflow_id,
            "node_id": nd.node_id,
            "memo_key": mk,
            "args_transport": transport,
            "input_artifact_ids": input_ids,
        }

        if nd.executor == "engine":
            out = await workflow.execute_activity(
                "run_engine_node",
                args=[req],
                start_to_close_timeout=_NODE_TIMEOUT,
                retry_policy=_ENGINE_RETRY,
            )
            ref = out["ref"]
            (self.executed if out["fresh"] else self.memo_hits).append(nd.node_id)
            if not out["fresh"]:
                await self._attach(ref)
            return ArtifactHandle(ref)

        # human node: ensure the (deduped) task exists, then wait for the answer
        self.human_waits.append(nd.node_id)
        await workflow.execute_activity(
            "ensure_human_task",
            args=[req],
            start_to_close_timeout=timedelta(seconds=60),
            retry_policy=RetryPolicy(
                maximum_attempts=10, non_retryable_error_types=["NodeError"]
            ),
        )
        delay = 1.0
        while ref is None:
            await workflow.sleep(delay)
            delay = min(delay * 2, 30.0)
            ref = await workflow.execute_activity(
                "memo_lookup",
                args=[self.engagement_id, mk],
                start_to_close_timeout=_SHORT,
            )
        self.executed.append(nd.node_id)
        await self._attach(ref)
        return ArtifactHandle(ref)

    async def _attach(self, ref: dict) -> None:
        await workflow.execute_activity(
            "attach_artifact",
            args=[self.workflow_run_id, ref["artifact_id"]],
            start_to_close_timeout=_SHORT,
        )

    def summary(self) -> dict:
        return {
            "workflow_run_id": self.workflow_run_id,
            "executed": self.executed,
            "memo_hits": self.memo_hits,
            "human_waits": self.human_waits,
        }

File: backend/src/temporal/Context.ts
import { ApplicationFailure } from '@temporalio/common';
import { proxyActivities, sleep } from '@temporalio/workflow';
import { ArtifactHandle } from '../domain/artifact/ArtifactHandle.js';
import type { ArtifactRef } from '../domain/artifact/ArtifactRef.js';
import { hashValue, memoKey } from '../domain/canonical/Canonical.js';
import type { JsonValue } from '../domain/json/JsonValue.js';
import type { NodeArgValue, NodeDef, Registry } from '../domain/registry/Registry.js';
import type { Acts } from './Activities.js';

// Ctx: the memoize-or-execute walk. SANDBOX code — imports limited to @temporalio/*, domain
// modules and type-only activity shapes; every DB/IO touch happens in an activity.

// RunInput/Summary/NodeRequest keep snake_case keys: they live in Temporal payloads and
// DB-adjacent transport, so the key spelling is wire contract.
export interface RunInput {
  engagement_id: number;
  workflow_run_id: number;
  workflow_id: string;
  declared_kinds: string[];
  attachments: ArtifactRef[];
}

export interface Summary {
  workflow_run_id: number;
  executed: string[];
  memo_hits: string[];
  human_waits: string[];
}

export type TransportValue =
  | JsonValue
  | { __artifact__: ArtifactRef }
  | TransportValue[]
  | { [key: string]: TransportValue };

export interface NodeRequest {
  engagement_id: number;
  workflow_run_id: number;
  workflow_id: string;
  node_id: string;
  memo_key: string;
  args_transport: TransportValue;
  input_artifact_ids: number[];
}

interface EncodedArgs {
  hashForm: JsonValue;
  transport: TransportValue;
  inputArtifactIds: number[];
}

const compareStrings = (a: string, b: string): number => {
  if (a < b) {
    return -1;
  }
  if (a > b) {
    return 1;
  }
  return 0;
};

// Non-empty all-ArtifactHandle lists are sorted by content hash (canonical rule 6); mixed or
// non-artifact lists keep caller order.
function sortedIfAllArtifacts(items: readonly NodeArgValue[]): readonly NodeArgValue[] {
  const handles: ArtifactHandle[] = [];
  for (const item of items) {
    if (!(item instanceof ArtifactHandle)) {
      return items;
    }
    handles.push(item);
  }
  if (handles.length === 0) {
    return items;
  }
  return handles.sort((a, b) => compareStrings(a.hash, b.hash));
}

// Three parallel encodings of a node argument value: the hash form ({$artifact: <hash>}), the
// transport form ({__artifact__: ref}) sent to activities, and the input artifact ids for the ledger.
function encodeArgs(value: NodeArgValue): EncodedArgs {
  if (value instanceof ArtifactHandle) {
    return {
      hashForm: { $artifact: value.hash },
      transport: { __artifact__: { ...value.ref } },
      inputArtifactIds: [value.artifactId],
    };
  }
  if (Array.isArray(value)) {
    const encoded = sortedIfAllArtifacts(value).map(encodeArgs);
    return {
      hashForm: encoded.map((e) => e.hashForm),
      transport: encoded.map((e) => e.transport),
      inputArtifactIds: encoded.flatMap((e) => e.inputArtifactIds),
    };
  }
  if (value !== null && typeof value === 'object') {
    const hashForm: { [key: string]: JsonValue } = {};
    const transport: { [key: string]: TransportValue } = {};
    const inputArtifactIds: number[] = [];
    for (const [k, v] of Object.entries(value)) {
      const e = encodeArgs(v);
      hashForm[k] = e.hashForm;
      transport[k] = e.transport;
      inputArtifactIds.push(...e.inputArtifactIds);
    }
    return { hashForm, transport, inputArtifactIds };
  }
  return { hashForm: value, transport: value, inputArtifactIds: [] };
}

// Collect every ArtifactHandle inside an argument value — single handles, arrays, and nested
// containers — so the inputKinds check cannot be smuggled past by wrapping.
function collectHandles(value: NodeArgValue, out: ArtifactHandle[]): void {
  if (value instanceof ArtifactHandle) {
    out.push(value);
    return;
  }
  if (Array.isArray(value)) {
    for (const item of value) {
      collectHandles(item, out);
    }
    return;
  }
  if (value !== null && typeof value === 'object') {
    for (const item of Object.values(value)) {
      collectHandles(item, out);
    }
  }
}

// Declared input ports, enforced before anything is hashed: a param mapped to a kind accepts only
// artifacts of that kind (single or list); a scalar param (null) accepts no artifacts. Absent
// params (nulled by the caller) pass vacuously. Exported for unit tests — Ctx.node's happy path
// requires a live activity proxy, but this guard is pure.
export function enforceInputKinds(nd: NodeDef, argMap: Record<string, NodeArgValue>): void {
  for (const p of nd.paramNames) {
    const expected = nd.inputKinds[p] ?? null;
    const handles: ArtifactHandle[] = [];
    collectHandles(argMap[p] ?? null, handles);
    if (expected === null) {
      if (handles.length > 0) {
        throw ApplicationFailure.nonRetryable(
          `node ${nd.nodeId}: param '${p}' is declared scalar but received an artifact of kind '${handles[0]?.kind ?? ''}'`
        );
      }
      continue;
    }
    for (const h of handles) {
      if (h.kind !== expected) {
        throw ApplicationFailure.nonRetryable(
          `node ${nd.nodeId}: param '${p}' expects kind '${expected}' but received '${h.kind}'`
        );
      }
    }
  }
}

const short = proxyActivities<Acts>({ startToCloseTimeout: '30s' });
const engineNode = proxyActivities<Acts>({
  startToCloseTimeout: '120s',
  retry: { maximumAttempts: 5, nonRetryableErrorTypes: ['NodeError'] },
});
const humanNode = proxyActivities<Acts>({
  startToCloseTimeout: '60s',
  retry: { maximumAttempts: 10, nonRetryableErrorTypes: ['NodeError'] },
});

export class Ctx {
  readonly engagementId: number;
  readonly workflowRunId: number;
  readonly workflowId: string;
  private readonly declared: Set<string>;
  private readonly attachments: ArtifactHandle[];
  private readonly registry: Registry;
  private readonly executed: string[] = [];
  private readonly memoHits: string[] = [];
  private readonly humanWaits: string[] = [];

  constructor(inp: RunInput, registry: Registry) {
    this.engagementId = inp.engagement_id;
    this.workflowRunId = inp.workflow_run_id;
    this.workflowId = inp.workflow_id;
    this.declared = new Set(inp.declared_kinds);
    this.attachments = inp.attachments.map((ref) => new ArtifactHandle(ref));
    this.registry = registry;
  }

  // User-sourced snapshot only (invariant I7), sorted by content hash.
  attached(kind: string): ArtifactHandle[] {
    if (!this.declared.has(kind)) {
      throw ApplicationFailure.nonRetryable(`kind '${kind}' is not declared by workflow '${this.workflowId}'`);
    }
    return this.attachments.filter((h) => h.kind === kind).sort((a, b) => compareStrings(a.hash, b.hash));
  }

  attachedOne(kind: string): ArtifactHandle {
    const items = this.attached(kind);
    const first = items[0];
    if (items.length !== 1 || first === undefined) {
      throw ApplicationFailure.nonRetryable(`expected exactly one '${kind}' attachment, found ${items.length}`);
    }
    return first;
  }

  attachedOneOrNone(kind: string): ArtifactHandle | null {
    const items = this.attached(kind);
    if (items.length > 1) {
      throw ApplicationFailure.nonRetryable(`expected at most one '${kind}' attachment, found ${items.length}`);
    }
    const first = items[0];
    return first === undefined ? null : first;
  }

  // Both spellings are public workflow-author API; the aliases delegate so overriding attached* covers both.
  userSupplied(kind: string): ArtifactHandle[] {
    return this.attached(kind);
  }

  userSuppliedOne(kind: string): ArtifactHandle {
    return this.attachedOne(kind);
  }

  userSuppliedOneOrNone(kind: string): ArtifactHandle | null {
    return this.attachedOneOrNone(kind);
  }

  // THE walk: input_hash over the canonical argument map, memo_key = H(node_id ':' input_hash),
  // lookup -> hit: reuse; miss: execute (engine activity, or human-task workflow + poll).
  async node(def: NodeDef, args: Record<string, NodeArgValue>): Promise<ArtifactHandle> {
    const nd = this.registry.tryNodeForWorkflow(this.workflowId, def.nodeId);
    if (nd === undefined) {
      throw ApplicationFailure.nonRetryable(
        `node '${def.nodeId}' is not registered for workflow '${this.workflowId}' (missing from its nodes list?)`
      );
    }
    const argMap: Record<string, NodeArgValue> = {};
    for (const [k, v] of Object.entries(args)) {
      if (!nd.paramNames.includes(k)) {
        throw ApplicationFailure.nonRetryable(`unknown parameter '${k}' for node ${nd.nodeId}`);
      }
      if (v !== undefined) {
        argMap[k] = v;
      }
    }
    // Absent optional input: explicit null (rule 7) — the memo key covers ALL parameters, always.
    // Own-property semantics (`in` would see Object.prototype names) and undefined-as-absent, so
    // omitted, undefined, and explicit-null calls all produce the SAME memo key.
    for (const p of nd.paramNames) {
      if (!Object.hasOwn(argMap, p)) {
        argMap[p] = null;
      }
    }
    enforceInputKinds(nd, argMap);
    const { hashForm, transport, inputArtifactIds } = encodeArgs(argMap);
    const mk = memoKey(nd.nodeId, hashValue(hashForm));

    const hit = await short.memo_lookup(this.engagementId, mk);
    if (hit !== null) {
      this.memoHits.push(nd.nodeId);
      await this.attach(hit);
      return new ArtifactHandle(hit);
    }

    const req: NodeRequest = {
      engagement_id: this.engagementId,
      workflow_run_id: this.workflowRunId,
      workflow_id: this.workflowId,
      node_id: nd.nodeId,
      memo_key: mk,
      args_transport: transport,
      input_artifact_ids: inputArtifactIds,
    };

    if (nd.executor === 'engine') {
      const out = await engineNode.run_engine_node(req);
      (out.fresh ? this.executed : this.memoHits).push(nd.nodeId);
      if (!out.fresh) {
        // The fresh path's attachment already happened inside the completion tx.
        await this.attach(out.ref);
      }
      return new ArtifactHandle(out.ref);
    }

    // Human path: park the question, then poll the memo with capped exponential backoff
    // (sleep first; 1s -> x2 -> 30s cap).
    this.humanWaits.push(nd.nodeId);
    await humanNode.ensure_human_task(req);
    let delay = 1;
    let ref: ArtifactRef | null = null;
    while (ref === null) {
      await sleep(delay * 1000);
      delay = Math.min(delay * 2, 30);
      ref = await short.memo_lookup(this.engagementId, mk);
    }
    this.executed.push(nd.nodeId);
    await this.attach(ref);
    return new ArtifactHandle(ref);
  }

  summary(): Summary {
    return {
      workflow_run_id: this.workflowRunId,
      executed: this.executed,
      memo_hits: this.memoHits,
      human_waits: this.humanWaits,
    };
  }

  private async attach(ref: ArtifactRef): Promise<void> {
    await short.attach_artifact(this.workflowRunId, ref.artifact_id);
  }
}
